components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt

```diff
@@ -0,0 +1,382 @@
+                            backendTerminalManager.createNewSharedTerminal(
+                                    "GitpodTerminal",
+                                    "Terminal"
+                            )
+                        } else {
+                            val aliasToTerminalMap:
+                                    MutableMap<String, TerminalOuterClass.Terminal> =
+                                    mutableMapOf()
+
+                            for (terminal in terminalsList) {
+                                val terminalAlias = terminal.alias
+                                aliasToTerminalMap[terminalAlias] = terminal
+                            }
+
+                            for (task in tasksList) {
+                                val terminalAlias = task.terminal
+                                val terminal = aliasToTerminalMap[terminalAlias]
+
+                                if (terminal != null) {
+                                    createSharedTerminal(terminal)
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private fun afterTerminalToolWindowGetsRegistered(action: () -> Unit) {
+        debug("Waiting for TerminalToolWindow to be registered...")
+        val toolWindowManagerListener =
+                object : ToolWindowManagerListener {
+                    override fun toolWindowsRegistered(
+                            ids: MutableList<String>,
+                            toolWindowManager: ToolWindowManager
+                    ) {
+                        if (ids.contains(TerminalToolWindowFactory.TOOL_WINDOW_ID)) {
+                            debug("TerminalToolWindow got registered!")
+                            action()
+                        }
+                    }
+                }
+
+        project.messageBus
+                .connect()
+                .subscribe(ToolWindowManagerListener.TOPIC, toolWindowManagerListener)
+    }
+
+    private fun withSupervisorTasksList(action: (tasksList: List<TaskStatus>) -> Unit) {
+        val taskStatusRequest = TasksStatusRequest.newBuilder().setObserve(true).build()
+
+        val taskStatusResponseObserver =
+                object : StreamObserver<TasksStatusResponse> {
+                    override fun onNext(response: TasksStatusResponse) {
+                        debug("Received task list: ${response.tasksList}")
+
+                        var hasOpenedAllTasks = true
+
+                        response.tasksList.forEach { task ->
+                            if (task.state === TaskState.opening) {
+                                hasOpenedAllTasks = false
+                            }
+                        }
+
+                        if (hasOpenedAllTasks) {
+                            this.onCompleted()
+                            action(response.tasksList)
```

Comment:
Should not we cancel streaming here?

Reply:
Calling `this.onCompleted()` ends the streaming, as I understand.